### PR TITLE
[RFC] package: add Shortcut Forwarding Engine driver

### PR DESCRIPTION
**Pull request by dissent1**
Add an opensource Qualcomm Shortcut FE driver.

The driver consists of:
- sfe (core modules): shortcut-fe (ipv4) and shortcut-fe-ipv6 - modules that actually do the offloading
- shortcut-fe-cm - simple shortcut-fe connection manager - a module that creates offloading rules and passes to sfe
- fast-classifier - advanced connection manager - a module that creates offloading rules and passes to sfe. Comparing to sfe-cm has additional checks, connection statistics and creates an offloading rule only when a connection hits 128 packets flowed (can be adjusted through /sys/fast_classifier/offload_at_pkts)
- userspace-fast-classifier - a fast-classifier user-space tool to manually add connections one want to offload
Both sfe-cm and fast-classifier get information about connections through netfilter. And after some checks it creates an offloading rule that is passed to sfe. After that sfe marks a connection as fast-forwarded and routes the flow.

Kernel parses this mark and bypasses its network stack taps for a specific connection in favor of sfe.

There are 3 patches added with this commit:
950 - adds a hook into network stack, sets no tcp window check to 0 (according to sfe readme).
951 - because fast-classifier also supports bridge offloading (echo 1 > /sys/fast_classifier/skip_to_bridge_ingress) and offloaded connections do not naturally update bridge statistics, fast-classifier will do that.
952 - allow multiple listeners for conntrack events, it is requied for netlink and sfe coexistance (for example)

All these patches are filled with ifdefs so the sfe code path is engaged only when these modules are compiled. It allows to keep kernel code untouched in case of no sfe compiled in.

Because sfe-cm and fast-classifier are competing modules, It's recommended to keep loaded only one of them. If both are loaded then in a race condition sfe-cm always comes first.

The driver is a merger of https://source.codeaurora.org/quic/qsdk/oss/lklm/shortcut-fe/log/?h=release/endive_cc with https://chromium.googlesource.com/chromiumos/third_party/kernel/+/chromeos-3.18/drivers/net/ethernet/qualcomm/sfe/

I've also added missing function to update udp statistics by fast-classifier and some more small fixes.

Patches are taken from QSDK and that chromeos-3.18 repo and a bit adjusted.

### CODE REVIEW

**Comment by dissent1** [comment]
Forgot to mention, SQM ingress is is not supported.
Benchmarking can be found in this thread https://forum.lede-project.org/t/qualcomm-fast-path-for-lede/4582

**Comment by dissent1** [comment]
To resolve the confusion it should be noted, that this driver is architecture agnostic and does not rely on hardware engines.

**Comment by dissent1** [comment]
Fixed a bug that pops when compiling with highest debug level
https://github.com/dissent1/r7800/commit/bd2a683e28367e850f4f83f50c81ff9a73c92af6

**Comment by dissent1** [comment]
SFE does not break sqm anymore. Put patches into hack-4.x dir.
Update the commit with following changes:
https://github.com/dissent1/r7800/commit/0468857c2c021fa34e42289c4321b90d8821a46d
https://github.com/dissent1/r7800/commit/bc187a8dda624579a6fd16197100fc75ea1bc552
https://github.com/dissent1/r7800/commit/e5e150a931bc862b2dd11aaac619e9c2c1d0059e

**Comment by philjohn** [comment]
Since those commits the build fails, due to the patches being in hack-4.* instead of patches-4.*

**Comment by dissent1** [comment]
That's actually correct. If you are compiling for stable 17.01 branch then you should manually copy kernel patches into patches-4.4 dir

**Comment by philjohn** [comment]
Yeah, that's what I ended up doing, thanks for confirming though!

I've got the fastpath patches applied to two different routers running custom builds of the 17.01 branch, both showing the different routes using fastpath in /sys/fashpath/debug_info so it looks like it's working as intended, and hasn't broken anything else (not currently using SQM)

**Comment by ypjalt** [comment]
MWAN3 doesn't work

**Comment by thagabe** [comment]
Noob here! Can I just pull this into a branch, commit, then build for my router (ARM)? Or do i need to go through the menuconfig and enable stuff? Right now I have the branch with the patches merged and building but nothing on my menuconfig changed. Please let me know if i can help in anyway (tests or otherwise)

**Comment by MartB** [comment]
👍 Nice work

**Comment by tapper82** [comment]
Any news on this? Is it working with SQM now?

**Comment by davidzodelin** [comment]
@tapper82 yes, working with SQM.

**Comment by montvid** [comment]
Would be nice to have it merged because of speed, cpu usage improvements. I would like to test this commit but first I have to learn how to build it for tp-link 1043nd v2.

**Comment by thagabe** [comment]
I tried an image with this patched merged. Worsened the through put and stability. I would like to help get this working right on arm devices!

**Comment by philjohn** [comment]
@thagabe I've got it running successfully on an R7800, which is ARM without issues. I'll try and paste some enabled/disabled iperf3 benches when I get home.

Has been stable for over a week now, so not seeing any instability here.

**Comment by philjohn** [comment]
@montvid you could always just take gwlim's precompiled fastpath build. I've got a 1043ND V1 in my network and have built a custom image with this patch applied, so it's not particularly difficult - contact me offline and I can provide my build env.

**Comment by montvid** [comment]
Thanks for the offer philjohn. I learned how to use git and patch LEDE so I tried to test the patch on 1043nd v2 connected two computers via lan. I did not want to reflash firmware so I just removed the packages, rebooted and tested. It seems either the removal of packages does not remove the patch or the performance is the same with or without the patch/packages. I guess more tests are needed with other hardware.
iperf3 -c 192.168.1.142 -i 1 -t 100 -V
Packages removed:
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-100.00 sec  10.9 GBytes   933 Mbits/sec    0             sender
[  4]   0.00-100.00 sec  10.9 GBytes   933 Mbits/sec                  receiver
CPU Utilization: local/sender 5.2% (0.1%u/5.1%s), remote/receiver 5.3% (0.4%u/4.9%s)
Packages installed:
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-100.00 sec  10.9 GBytes   934 Mbits/sec    0             sender
[  4]   0.00-100.00 sec  10.9 GBytes   933 Mbits/sec                  receiver
CPU Utilization: local/sender 5.2% (0.1%u/5.1%s), remote/receiver 3.4% (0.2%u/3.1%s)

**Comment by haruhitamako** [comment]
I put the patches into the patches-4.4 when I make 17.01 ,Will the SQM be broken again？

**Comment by ezar** [comment]
Someone can help me to add this PR to current environment?

**Comment by danielpetre** [comment]
20 days and this awesome patchset still isn't commited to LEDE..?

**Comment by ezar** [comment]
Commit please!

**Comment by jeolivera** [comment]
SQM (cake and layer_cake.qos) with additional flags "mpu 64 nat dual-dst/srchost" seems to works alongside this. Also checked if it was even working by cat /sys/fast_classifier/debug_info.

**Comment by Mushoz** [comment]
This is probably one of the most sought-after enhancements of LEDE. Is there anybody working on reviewing the code? Would love to see this committed.

**Comment by commodo** [comment]
@dissent1 
So, I'll add a bit of initial set of thoughts [ that could count as an initial review ].

Can we change from adding all the source files into the repo and reference the git repo from codeaurora.org ?
If possible, what I would do is define a section in the main Makefile like:
```
PKG_SOURCE_URL:=https://source.codeaurora.org/quic/qsdk/oss/lklm/shortcut-fe
PKG_SOURCE_PROTO:=git
PKG_SOURCE_DATE:=2017-06-14
PKG_SOURCE_VERSION:=25e41dabed2cda9a95dc95e3383cf2a617f88de7
PKG_MAINTAINER:=Pavel Kubelun <be.dissent@gmail.com>
PKG_LICENSE:=<license used here>
```

FWIW: I checked that these values are correct (based on your link):
```
PKG_SOURCE_DATE:=2017-06-14
PKG_SOURCE_VERSION:=25e41dabed2cda9a95dc95e3383cf2a617f88de7
```

And then all patches from https://chromium.googlesource.com/chromiumos/third_party/kernel/+/chromeos-3.18/drivers/net/ethernet/qualcomm/sfe/
could be turned into file-patches and put into the `patches` folder.
If possible, I would make each patch from this repo, a singular commit, referencing a web-link to the commit in that repo.

Since the kernel patches are only 3, what I would do, is make each kernel patch, it's own commit, and reference a web-link to each commit from where they were obtained.
And then the config-4.4 & config-4.9 could be a single commit.

From a reviewer stand-point, this PR/patch, is very difficult to review [I mostly do reviews at work and it takes some energy to do].
But if you manage to remove all the source files, and just download from the source from `codeaurora.org`, and add patch files [in the `patches` folder], it would simplify things a great deal.
Also, because it reduces the chance/risk that any of these files were copied/patched wrongly [ the assumption here is that human error is very possible ]

`PKG_LICENSE` [ and `PKG_LICENSE_FILES` fields ] are very much needed;
The legal aspects are important for this kind of project [as LEDE & OpenWrt are].
I hope that there aren't any legal aspects preventing the SFE engine to be packaged into LEDE :)

`PKG_MAINTAINER` is preferred, since you are implicitly taking ownership of this module [in LEDE], by committing this change.

If you need clarification or help in organizing this, feel free to ping me.
I can probably alloc some time this week to work with you into splitting this into a PR that is more manage-able for review.
That is the best that I can do, since I'm just an occasional contributor.

**Comment by tapper82** [comment]
I am happy to see that people are still talking about this as I thought it was a ded duck! @commodo  is rite about this being the biggest thing for LEDE /OpenWRT for a long time.

**Comment by jack338c** [comment]
@dissent1 

Dear dissent
Could you please tell me How should I apply this patch? if I am compiling for stable 17.01 branch

is that like following steps?

1.wget https://patch-diff.githubusercontent.com/raw/lede-project/source/pull/1269.patch into"lede source tree" or to "kernel patches into patches-4.4"
2.in source tree do: git apply --ignore-space-change --ignore-whitespace 1269.patch
3.then in menuconfig  Kernel Modules > Network Support > kmod-fast-classifier and kmod-shortcut-fe (not kmod-shortcut-fe-cm)

please  hele me 
thank you

**Comment by montvid** [comment]
@jack338c I just merged the patch with the master branch. I could not find out how to do it on the stable but some people in this pull did it.
@dissent1 I was trying to make sense of the debugging information provided by the driver

> cat /sys/fast_classifier/debug_info
cat /sys/fast_classifier/exceptions

 but it is very hard to understand. I would dream to have a easy to understand output implemented showing what is happening and if there are any problems with the driver. Like how can I tell if the driver is working as it should if I need to ask you instead of understanding the debug info myself.

**Comment by jack338c** [comment]
@montvid 
thanks for your reply
you can try speedtest to verify the patch whether work or not
if it works you will feel the speed up obviously while dowloading or uploading

**Comment by jack338c** [comment]
Dear philjohn 
you said: Kernel Modules > Network Support > kmod-fast-classifier and kmod-shortcut-fe (not kmod-shortcut-fe-cm)
but dissent1 said: Because sfe-cm and fast-classifier are competing modules, It's recommended to keep loaded only one of them. If both are loaded then in a race condition sfe-cm always comes first.

so,I suppose I should select only : Kernel Modules > Network Support >kmod-fast-classifier 
or only  Kernel Modules > Network Support >kmod-shortcut-fe-cm
is that right?

thank you

**Comment by philjohn** [comment]
@jack338c nope, kmod-fast-classifier and kmod-shortcut-fe is correct.

kmod-shortcut-fe-cm is a simple connection manager, fast-classifier is more advanced and is used in its place.

@dissent1 I'm playing around with the QSDK nss support, I've tried to add sfe_drv.* and getting 


/lede/R7800/source/build_dir/target-arm_cortex-a15+neon-vfpv4_musl-1.1.16_eabi/linux-ipq806x/shortcut-fe/sfe_drv.c:1262:9: error: 'athrs_fast_nat_recv' undeclared (first use in this function)

any idea why?

**Comment by jack338c** [comment]
Ok, I got it thanks

**Comment by montvid** [comment]
So maybe let's disable kmod-shortcut-fe-cm so people would not ask at all which one to choose? :)

**Comment by philjohn** [comment]
@montvid +1

I've got the shortcut-fe-drv compiling now, which is required for Qualcomm NSS hardware acceleration support, simple change from athrs_fast* to fast* in sfe_drv.c, might be worth adding to this pull request to support that going forwards.

**Comment by ezar** [comment]
+1

File: package/kernel/shortcut-fe/src/fast-classifier.c [hunk 1]
```diff
@@ -0,0 +1,1892 @@
...
+		struct rtnl_link_stats64 nlstats;
+		nlstats.tx_packets = 0;
+		nlstats.tx_bytes = 0;
+
+		if (sis->src_dev && IFF_EBRIDGE &&
+		    (sis->src_new_packet_count || sis->src_new_byte_count)) {
+			nlstats.rx_packets = sis->src_new_packet_count;
+			nlstats.rx_bytes = sis->src_new_byte_count;
+			spin_lock_bh(&sfe_connections_lock);
+			br_dev_update_stats(sis->src_dev, &nlstats);
```
Comment (MartB) [inline package/kernel/shortcut-fe/src/fast-classifier.c]: If CONFIG_BRIDGE is not defined this will error out while loading (not while compiling) the module as the symbol does not exist.

This block could be wrapped with `#if IS_ENABLED(CONFIG_BRIDGE)`, but it probably would be cleaner to remove everything related to bridges if its disabled. Not sure how far away we go from upstream with that though.

**Comment by chros73** [comment]
@devs: Take a look at [this test](https://forum.lede-project.org/t/qualcomm-fast-path-for-lede/4582/479?u=chros).

@philjohn: what is NSS hardware acceleration for? Thanks

File: target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+ /* deliver cached events and clear cache entry - must be called with locally
+  * disabled softirqs */
++#ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
++void nf_ct_deliver_cached_events(struct nf_conn *ct)
++{
++	unsigned long events, missed;
++	struct nf_conntrack_ecache *e;
++	struct nf_ct_event item;
++	struct net *net = nf_ct_net(ct);
++	int ret = 0;
```
Comment (MartB) [inline target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch]: This can be removed as its value is alway 0, you can also remove all the ifs depending on that value.

File: target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
++
++	item.ct = ct;
++	item.portid = 0;
++	item.report = 0;
++
++	atomic_notifier_call_chain(&net->ct.nf_conntrack_chain,
++			events | missed,
++			&item);
++
++	if (likely(ret >= 0 && !missed))
```
Comment (MartB) [inline target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch]: remove the ret >= 0 its always true

File: target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
++		return;
++
++	spin_lock_bh(&ct->lock);
++	if (ret < 0)
```
Comment (MartB) [inline target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch]: remove this entire condition and replace it with the code from the else block

**Comment by philjohn** [comment]
> @philjohn: what is NSS hardware acceleration for? Thanks

@chros73 the IPQ806X as 2 extra cores, based on UBI32, that act as a network acceleration subsystem and offloads processing of packets from the ARM cores. All of the code is available on the CodeAurora site, including a Linux 4.4.60 tree with the various bits and pieces needed, so I'm porting over patches for the needed hooks in the network code. It accelerates things including PPPoE and IPSEC using the hardware crypto routines built into the silicon.

File: target/linux/generic/hack-4.9/950-net-patch-linux-kernel-to-support-shortcut-fe.patch [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+diff --git a/net/netfilter/nf_conntrack_proto_tcp.c b/net/netfilter/nf_conntrack_proto_tcp.c
+index f24b626..4581481 100644
+--- a/net/netfilter/nf_conntrack_proto_tcp.c
++++ b/net/netfilter/nf_conntrack_proto_tcp.c
+@@ -34,12 +34,22 @@
+ #include <net/netfilter/ipv6/nf_conntrack_ipv6.h>
+ 
+ /* Do not check the TCP window for incoming packets  */
++#ifdef CONFIG_SHORTCUT_FE
++int nf_ct_tcp_no_window_check __read_mostly = 0;
```
Comment (MartB) [inline target/linux/generic/hack-4.9/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: I know this is in the readme but the chrome os devs decided to remove it again.
https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/271601

Leaving that one up for a reviewer that is more involved with the project and this matter.

File: target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
++EXPORT_SYMBOL_GPL(nf_ct_tcp_no_window_check);
++#else
+ static int nf_ct_tcp_no_window_check __read_mostly = 1;
++#endif
+ 
+ /* "Be conservative in what you do,
+     be liberal in what you accept from others."
+     If it's non-zero, we mark only out of window RST segments as INVALID. */
++#ifdef CONFIG_SHORTCUT_FE
++int nf_ct_tcp_be_liberal __read_mostly;
```
Comment (MartB) [inline target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: Why is this needed, care to explain ?
If it was not done by you do you have the origin of that patch ?

File: target/linux/generic/hack-4.9/950-net-patch-linux-kernel-to-support-shortcut-fe.patch (hunk 5 above)
Comment (dissent1) [reply target/linux/generic/hack-4.9/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: I think you are a bit misunderstanding this part. We are turning the window check back on in kernel if the SFE is compiled in, it’s similar to reverting the patch that introduces this possibility.
As a side note: you are pointing to the wrong tree, where SFE has been pushed into chromeos - it’s chromeos-3.18.

File: target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch (hunk 6 above)
Comment (dissent1) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: This symbol along with tcp_no_window_check is used in both connection managers and trigger some additional checks if un/set.
It also originates from qsdk SFE repo, placed in patches folder.

Comment (dissent1) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: Actually I’ve just copy/pasted patches for k4.4 from QSDK. As you see I already have these points cleaned out in patch for k4.9. I’m actually not sure if there is a need for k4.4 here, as LEDE is switching off it in master branch and unlikely this package makes its way into stable 17.01

Comment (dissent1) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: ops, wrong point

File: target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch (hunk 4 above)
Comment (dissent1) [reply target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch]: Actually I’ve just copy/pasted patches for k4.4 from QSDK. As you see I already have these points cleaned out in patch for k4.9. I’m actually not sure if there is a need for k4.4 here, as LEDE is switching off it in master branch and unlikely this package makes its way into stable 17.01

File: target/linux/generic/hack-4.9/950-net-patch-linux-kernel-to-support-shortcut-fe.patch (hunk 5 above)
Comment (MartB) [reply target/linux/generic/hack-4.9/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: Yep just found the commit that reintroduced it exclusively for sfe, my bad.

(https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/380438)

File: target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch (hunk 4 above)
Comment (MartB) [reply target/linux/generic/hack-4.4/952-net-conntrack-events-support-multiple-registrant.patch]: Uhm atleast for ar71xx it comes in pretty handy.
Still stuck on 4.4 there.

But i guess you are right just wanted to point that out.

File: target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch (hunk 6 above)
Comment (MartB) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: Thanks!

Comment (jack338c) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: Dear dissent1
Could you tell me what 's the different betweent yours and glw's 
who's respositoryof fast patch is more faster?
thank you

Comment (philjohn) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: @jack338c I'm not dissent1, but from what I can see between the two, @dissent1 has stayed closer to the upstream project in QSDK whereas gwlim's patchset includes a heap of compile-time optimization for the mips 24 and 74k arch.

Comment (jack338c) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: @philjohn 
thanks for patience and help
I test both with my netgear wndr4300 74k and found gwlim's is more faster,especially on playing videos ,I can see the buffer is more quicker on gwlim's
but I also want to use dissent1,it's simple for compling

Comment (jack338c) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: hi @philjohn 
I just bought a R7800
if I want compile a firmware from https://github.com/dissent1/r7800/tree/bd2a683e28367e850f4f83f50c81ff9a73c92af6

how to git clone the repository with shortcut-fe
if I git clone https://github.com/dissent1/r7800.git ,there will be no shortcut-fe.
please help me ,thank you

Comment (philjohn) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: add .patch to the end of this URL, so the end is 1269.patch, then apply it to the source tree.

Failing that, have a look into hnyman's builds in the forum as I believe one of his now has fastpath support.

Comment (jack338c) [reply target/linux/generic/hack-4.4/950-net-patch-linux-kernel-to-support-shortcut-fe.patch]: sorry what is "this URL"?
Could I just download a zipfile from https://github.com/dissent1/r7800/tree/bd2a683e28367e850f4f83f50c81ff9a73c92af6 ,then exact it for compile?
thanks for fast reply

**Comment by haruhitamako** [comment]
i try to make it in lede-17.01.2,but there are song error:
/home/haruhi/source-17.01.2/build_dir/target-mipsel_24kc_musl-1.1.16/linux-ramips_mt7620/shortcut-fe/sfe_ipv4.c:1366:5: error: 'struct sk_buff' has no member named 'fast_forwarded'
  skb->fast_forwarded = 1;
     ^
/home/haruhi/source-17.01.2/build_dir/target-mipsel_24kc_musl-1.1.16/linux-ramips_mt7620/shortcut-fe/sfe_ipv4.c: In function 'sfe_ipv4_recv_tcp':
/home/haruhi/source-17.01.2/build_dir/target-mipsel_24kc_musl-1.1.16/linux-ramips_mt7620/shortcut-fe/sfe_ipv4.c:1909:5: error: 'struct sk_buff' has no member named 'fast_forwarded'
  skb->fast_forwarded = 1;

**Comment by MOZGIII** [comment]
Since master now uses linux 4.9, should we expect cleanup in this PR?

**Comment by thagabe** [comment]
@hauke @blogic @jow- This has been working great for many people on the forum. A decision should be made or comments would be appreciated. Thanks

**Comment by Misiek304** [comment]
I've had a few serious problems with SQM (cake) and SFE. The SQM itself worked fine but more advanced options of cake didn't. I believe that SFE cripples cake's potential.

**Comment by thagabe** [comment]
@Misiek304 I have noticed this too, however (as my use case is a small network) i opted to disable cake in favor of SFE. Maybe SFE should disable cake if it is selected. At lease until a proper fix, if there is one, is hammered out.

**Comment by Misiek304** [comment]
Unfortunately I can't afford to disable SQM and cake in my case use. I depend on it too much so I opted out to get rid of SFE. I hope that in the future releases this issue will get resolved. 

For now I will satisfy on a clear warning notice that SQM and cake don't work properly with SFE active. I've seen many claims to the contrary on the forum and I think this needs to be addressed to at least avoid flood of issue threads about latency and bufferbload. I also think that not having a clear statement on this subject hurts entire SQM and cake projects since it greatly affects peoples reception about cake effectiveness.

The best case scenario would be to have ISP's implement cake in their cable modems.

**Comment by davidzodelin** [comment]
What are the advanced options of SQM Cake that in your case does not work? I've been using SFE and SQM Cake for 2 months in a real environment and the only problem I've noticed is that SFE does not work well with long-distance wireless links.

  Please, I need to know what are those advanced options that do not work for you because in my case I also depend on them.

**Comment by Misiek304** [comment]
option iqdisc_opts 'nat dual-dsthost diffserv4 rtt 300ms mpu 64'
	option eqdisc_opts 'nat dual-srchost diffserv4 rtt 300ms mpu 64'

Those settings did not work for sure. As soon as I removed SFE those settings were activated.

**Comment by MOZGIII** [comment]
Afaik traffic that goes fast path is not affected by SQM. That would explain why doesn't work. Am I right?

**Comment by davidzodelin** [comment]
It's strange. I use the following features:

diffserv4 triple-isolate nat rtt 100.0ms noatm overhead 44 via-ethernet mpu 64

  And before my ISP removed the access points nearby that were 100 meters away everything worked perfectly. Now I have noticed that SFE goes wrong when I connect to an access point that is approximately 500 meters away. Are you sure you are using the SFE with the fixes that Dissent did for the ingress when using tc qdisc?

**Comment by Misiek304** [comment]
The `nat dual-srchost` didn't work for sure. I've tested this. It behaved like those options were non existent. As soon as I removed SFE, cake behaved well and I saw big improvement. SFE affected mostly upload performance.

**Comment by lucize** [comment]
something is strange, I have a cisco ip phone 7960 behind a router linked to the Call Manager by wireguard, and after adding the fastpath option to the router the calls can't be made by the first attempt, and is strange because I'm doing routing not NAT to other devices in VPN

**Comment by lucize** [comment]
@dissent1 this is really needed ?

```
 DEPENDS:=@IPV6
 FILES:=$(PKG_BUILD_DIR)/shortcut-fe.ko $(PKG_BUILD_DIR)/shortcut-fe-ipv6.ko
```
can't it be compiled without IPV6, maybe change a bit the logic about `shortcut-fe-ipv6.ko` ?

File: package/kernel/shortcut-fe/Makefile [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+PKG_NAME:=shortcut-fe
+PKG_RELEASE:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+define KernelPackage/shortcut-fe
+  SECTION:=kernel
+  CATEGORY:=Kernel modules
+  SUBMENU:=Network Support
+  DEPENDS:=@IPV6
```
Comment (lucize) [inline package/kernel/shortcut-fe/Makefile]: @dissent1 this is really needed ?
 can't it be compiled without IPV6 ?
maybe change a bit the logic about shortcut-fe-ipv6.ko ?

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

**Comment by Priyantha** [comment]
Hi @blogic 
How can we help, as users, to encourage the LEDE/OpenWRT developers to include this fantastic PR?

**Comment by blogic** [comment]
@Priyantha read the comments ... "With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days."

**Comment by jow-** [comment]
This particular change however is unlikely to land in OpenWrt as the project opts for the upstream fast path implementation instead.

**Comment by Priyantha** [comment]
Ah okay, there is an upstream fast path implementation? I wasn't aware of that 👍 when there is one, that would be great. It's a big improvement. So I am hoping it will be part of the new OpenWRT trunk in the near future.

**Comment by nbd168** [comment]
FYI, I've backported the upstream flow offload code to Linux 4.14 in my staging tree. I've also written an extra patch that integrates it with iptables (so no dependency on nftables anymore).
You can test it by building my staging tree, enabling kmod-ipt-flow, and adding a rule to the FORWARD chain with -j FLOWOFFLOAD

**Comment by lucize** [comment]
@nbd168 sorry for asking, can you give an example to see how it would need to be ?

**Comment by Mushoz** [comment]
I cannot find the mentioned kmod-ipt-flow in menuconfig. I do find these 3 modules, however:

kmod-ipt-offload
kmod-nf-flow
kmod-nft-offload

Why ones do I need?

**Comment by pexcn** [comment]
those packages will merge in master branch?

**Comment by LuKePicci** [comment]
Just for the records, the SFE code in this PR had also issues over IPSec VPNs. The issue troubleshooting is simple: as soon as a connection passing through the tunnel is picked for shortcutting it breaks, so you see the TCP layer often doing retransmission and, for instance, relatively large web pages load only partially.
